libcudacxx/include/cuda/std/__cuda/semaphore.h

```diff
@@ -38,8 +52,152 @@ class counting_semaphore : public _CUDA_VSTD::__semaphore_base<__least_max_value
   counting_semaphore& operator=(const counting_semaphore&) = delete;
 };
 
+template <thread_scope _Sco, ptrdiff_t __least_max_value = INT_MAX>
+struct fair_counting_semaphore
+{
+};
+
 template <thread_scope _Sco>
-using binary_semaphore = counting_semaphore<_Sco, 1>;
+struct binary_semaphore : fair_counting_semaphore<_Sco, 1>
+{
+public:
+    _LIBCUDACXX_INLINE_VISIBILITY static constexpr ptrdiff_t max() noexcept
+    {
+        return static_cast<ptrdiff_t>(1);
+    }
+
+    _LIBCUDACXX_INLINE_VISIBILITY bool try_acquire()
+    {
+        ptrdiff_t ctr = counter.load(cuda::memory_order_relaxed);
+        return try_acq_impl(ctr);
+    }
+
+    _LIBCUDACXX_INLINE_VISIBILITY void acquire()
+    {
+        ptrdiff_t ctr = counter.load(cuda::memory_order_relaxed);
+        while (!try_acq_impl(ctr))
+        {
+            counter.wait(ctr);
```

Comment:
I avoid calling `try_acquire` directly because I don't want to do a second, redundant load.